File: app/src/main/java/shashank/com/screenrecorder/EditVideoActivity.java
package shashank.com.screenrecorder;

import android.app.ProgressDialog;
import android.content.DialogInterface;
import android.net.Uri;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.NonNull;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.View;
import android.widget.ImageView;
import android.widget.SeekBar;
import android.widget.TextView;
import android.widget.VideoView;

import java.io.File;


/**
 * Created by shashankm on 23/03/17.
 */

public class EditVideoActivity extends AppCompatActivity implements CustomRange.RangeChangeListener,
    View.OnClickListener, EditVideoContract.Response {
    private static final String TAG = EditVideoActivity.class.getSimpleName();
    private Handler handler = new Handler();

    private VideoView video;
    private SeekBar seekBar;
    private CustomRange rangePicker;
    private ImageView playPause;
    private TextView startTime, endTime, videoCurrentTime;
    private final Handler hideHandler = new Handler();
    private String data;
    private ProgressDialog progressDialog;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_edit_video);

        data = getIntent().getStringExtra("data");
        video = (VideoView) findViewById(R.id.video);
        seekBar = (SeekBar) findViewById(R.id.seek_bar);
        rangePicker = (CustomRange) findViewById(R.id.video_range_picker);
        playPause = (ImageView) findViewById(R.id.play_pause);
        startTime = (TextView) findViewById(R.id.start_time);
        endTime = (TextView) findViewById(R.id.end_time);

        videoCurrentTime = (TextView) findViewById(R.id.video_current_time);
        View videoContainer = findViewById(R.id.video_container);
        View trimVideo = findViewById(R.id.trim_video);
        View back = findViewById(R.id.back);

        int duration = (int) getIntent().getLongExtra("duration", 0);

        int maxFileDuration = 1000 * 60 * 5; // 5 minutes
        if (duration > maxFileDuration) {
            AlertDialog.Builder builder = new AlertDialog.Builder(this)
                    .setTitle("Error!")
                    .setMessage("The file size is too large. It's gonna take forever to trim. Have mercy and pick a smaller file")
                    .setPositiveButton("Cool", new DialogInterface.OnClickListener() {
                        @Override
                        public void onClick(DialogInterface dialog, int which) {

                        }
                    });
            builder.show();
            return;
        }

        video.setVideoURI(Uri.parse(data));
        video.start();
        seekBar.setMax(duration);
        rangePicker.setMinValue(0);
        rangePicker.setMaxValue(duration);
        rangePicker.setRangeChangeListener(this);

        startTime.setText(AppUtil.INSTANCE.getMinsAndSecs(0f));
        endTime.setText(AppUtil.INSTANCE.getMinsAndSecs(duration));

        seekUpdation();
        hidePlayPause();

        seekBar.setOnSeekBarChangeListener(new SeekBar.OnSeekBarChangeListener() {
            @Override
            public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
                if (fromUser) {
                    video.seekTo(progress);
                    seekBar.setProgress(progress);
                    videoCurrentTime.setText(AppUtil.INSTANCE.getMinsAndSecs(progress));
                }
            }

            @Override
            public void onStartTrackingTouch(SeekBar seekBar) {

            }

            @Override
            public void onStopTrackingTouch(SeekBar seekBar) {

            }
        });

        videoContainer.setOnClickListener(this);
        trimVideo.setOnClickListener(this);
        back.setOnClickListener(this);
    }

    private void hidePlayPause() {
        hideHandler.postDelayed(new Runnable() {
            @Override
            public void run() {
                playPause.setVisibility(View.GONE);
            }
        }, 3000);
    }

    @Override
    public void onRangeChanged(float startValue, float endValue) {
        int startInt = (int) startValue;
        seekBar.setProgress(startInt);
        video.seekTo(startInt);
        video.pause();
        startTime.setText(AppUtil.INSTANCE.getMinsAndSecs(startValue));
        endTime.setText(AppUtil.INSTANCE.getMinsAndSecs(endValue));
    }

    private Runnable runnable = new Runnable() {
        @Override
        public void run() {
            seekUpdation();
        }
    };

    private void seekUpdation() {
        seekBar.setProgress(video.getCurrentPosition());
        videoCurrentTime.setText(AppUtil.INSTANCE.getMinsAndSecs(video.getCurrentPosition()));
        handler.postDelayed(runnable, 1000);
    }

    @Override
    public void onClick(View v) {
        switch (v.getId()) {
            case R.id.video_container:
                playPause.setVisibility(View.VISIBLE);
                if (video.isPlaying()) {
                    video.pause();
                    playPause.setImageResource(R.drawable.ic_play_arrow);
                } else {
                    video.start();
                    playPause.setImageResource(R.drawable.ic_pause);
                    hidePlayPause();
                }
                break;

            case R.id.trim_video:
                FfmpegUtil ffmpegUtil = new FfmpegUtil(this, this);

                String startTime = AppUtil.INSTANCE.getTime(rangePicker.getStartValue());
                String endTime = AppUtil.INSTANCE.getTime(rangePicker.getEndValue());
                ffmpegUtil.trimVideo(new File(Uri.parse(data).getPath()), startTime, endTime);
                break;

            case R.id.back:
                onBackPressed();
                break;
        }
    }

    @Override
    public void showProgress(@NonNull String title, @NonNull String message) {
        progressDialog = ProgressDialog.show(this, title, message);
        progressDialog.show();
    }

    @Override
    public void finishedSuccessFully(String path) {
        if (progressDialog != null) {
            progressDialog.dismiss();
        }

        new AlertDialog.Builder(this)
                .setTitle("Success!")
                .setMessage("Your file is successfully saved at " + path + " in your phone")
                .setPositiveButton("Great", new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {

                    }
                }).create().show();
    }

    @Override
    public void onFailure(@NonNull String message) {
        if (progressDialog != null) {
            progressDialog.dismiss();
        }
    }
}

File: app/src/main/java/shashank/com/screenrecorder/android/EditVideoActivity.kt
package shashank.com.screenrecorder.android

import android.app.ProgressDialog
import android.net.Uri
import android.os.Bundle
import android.os.Handler
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.view.View
import android.widget.SeekBar
import kotlinx.android.synthetic.main.activity_edit_video.*
import shashank.com.screenrecorder.R
import shashank.com.screenrecorder.helper.EditVideoContract
import shashank.com.screenrecorder.helper.FfmpegHelper
import shashank.com.screenrecorder.util.AppUtil
import shashank.com.screenrecorder.util.CustomRange
import java.io.File


/**
 * Created by shashankm on 23/03/17.
 */

class EditVideoActivity : AppCompatActivity(), CustomRange.RangeChangeListener, View.OnClickListener, EditVideoContract.Response {
    private val handler = Handler()

    private val hideHandler = Handler()
    private var data: String? = null
    private var progressDialog: ProgressDialog? = null

    public override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_edit_video)


        data = intent.getStringExtra("data")
        val duration = intent.getLongExtra("duration", 0).toInt()
        back.setOnClickListener(this)

        val maxFileDuration = 1000 * 60 * 5 // 5 minutes
        if (duration > maxFileDuration) {
            val builder = AlertDialog.Builder(this)
                    .setTitle("Error!")
                    .setMessage("The file size is too large. It's gonna take forever to trim. Have mercy and pick a smaller file")
                    .setPositiveButton("Cool") { _, _ -> }
            builder.show()
            return
        }

        video.setVideoURI(Uri.parse(data))
        video.start()
        seek_bar.max = duration
        video_range_picker.minValue = 0f
        video_range_picker.maxValue = duration.toFloat()
        video_range_picker.setRangeChangeListener(this)

        start_time!!.text = AppUtil.getMinsAndSecs(0f)
        end_time!!.text = AppUtil.getMinsAndSecs(duration.toFloat())

        seekUpdation()
        hidePlayPause()

        seek_bar.setOnSeekBarChangeListener(object : SeekBar.OnSeekBarChangeListener {
            override fun onProgressChanged(seekBar: SeekBar, progress: Int, fromUser: Boolean) {
                if (fromUser) {
                    video!!.seekTo(progress)
                    seekBar.progress = progress
                    video_current_time!!.text = AppUtil.getMinsAndSecs(progress.toFloat())
                }
            }

            override fun onStartTrackingTouch(seekBar: SeekBar) {

            }

            override fun onStopTrackingTouch(seekBar: SeekBar) {

            }
        })

        video_container.setOnClickListener(this)
        trim_video.setOnClickListener(this)
    }

    private fun hidePlayPause() {
        hideHandler.postDelayed({ play_pause!!.visibility = View.GONE }, 3000)
    }

    override fun onRangeChanged(startValue: Float, endValue: Float) {
        val startInt = startValue.toInt()
        seek_bar.progress = startInt
        video.seekTo(startInt)
        video.pause()
        start_time.text = AppUtil.getMinsAndSecs(startValue)
        end_time.text = AppUtil.getMinsAndSecs(endValue)
    }

    private val runnable = Runnable { seekUpdation() }

    private fun seekUpdation() {
        seek_bar.progress = video.currentPosition
        video_current_time.text = AppUtil.getMinsAndSecs(video.currentPosition.toFloat())
        handler.postDelayed(runnable, 1000)
    }

    override fun showBusy() {
        AlertDialog.Builder(this)
                .setTitle("Busy!")
                .setMessage("Please wait while we finish up with your earlier media transformation")
                .setPositiveButton("Ok", { _, _ ->  })
                .create().show()
    }

    override fun onClick(v: View) {
        when (v.id) {
            R.id.video_container -> {
                play_pause.visibility = View.VISIBLE
                if (video.isPlaying) {
                    video.pause()
                    play_pause.setImageResource(R.drawable.ic_play_arrow)
                } else {
                    video.start()
                    play_pause.setImageResource(R.drawable.ic_pause)
                    hidePlayPause()
                }
            }

            R.id.trim_video -> {
                val ffmpegUtil = FfmpegHelper(this, this)

                val startTime = AppUtil.getTime(video_range_picker.startValue)
                val endTime = AppUtil.getTime(video_range_picker.endValue)
                ffmpegUtil.trimVideo(File(Uri.parse(data).path), (video_range_picker.endValue - video_range_picker.startValue).toInt(),
                        startTime, endTime)

                AlertDialog.Builder(this)
                        .setMessage("Your media conversion has started, you can track its progress in the notification bar")
                        .setPositiveButton("Ok", { _, _ ->  })
                        .create().show()
            }

            R.id.back -> onBackPressed()
        }
    }

    override fun showProgress(title: String, message: String) {
        progressDialog = ProgressDialog.show(this, title, message)
        progressDialog!!.show()
    }

    override fun finishedSuccessFully(path: String?) {
        if (progressDialog != null) {
            progressDialog!!.dismiss()
        }

        AlertDialog.Builder(this)
                .setTitle("Success!")
                .setMessage("Your file is successfully saved at $path in your phone. You can also check it out in your gallery!")
                .setPositiveButton("Great") { _, _ -> }.create().show()
    }

    override fun onFailure(message: String) {
        if (progressDialog != null) {
            progressDialog!!.dismiss()
        }
    }
}
